traceevents/src/main/java/com/tomtom/kotlin/traceevents/Tracer.kt

```diff
@@ -522,20 +593,70 @@ class Tracer private constructor(
             }
         }
 
-        internal fun createLogMessage(traceEvent: TraceEvent): String {
-            return "[${traceEvent.dateTime.format(DateTimeFormatter.ISO_DATE_TIME)}] " +
+        internal fun registeredToString(item: Any?) =
+            if (item == null) {
+                "null"
+            } else {
+                // Try registered function first.
+                toStringRegistry[item::class.toString()]?.invoke(item)
+                    ?: if (item.javaClass.getMethod("toString").declaringClass == Any::class.java) {
+
+                        // Only default [toString] is available, from [Any].
+                        "${item.javaClass.simpleName}(...)"
+                    } else {
+
+                        // Use [toString] as defined in class itself.
+                        item.toString()
+                    }
+            }
+
+        /**
+         * Create a trace event message that can be logged.
+         */
+        internal fun createLogMessage(
+            traceEvent: TraceEvent,
+            includeTime: Boolean,
+            logStackTrace: Boolean,
+            includeOwnerClass: Boolean
+        ): String {
+            val sb = StringBuilder()
+            if (includeTime) {
+                sb.append("[${traceEvent.dateTime.format(DateTimeFormatter.ISO_DATE_TIME)}] ")
+            }
+            sb.append(
                 "${traceEvent.functionName}(${traceEvent.args.joinToString {
-                    toStringFromRegistered(it)
-                }}), from ${traceEvent.ownerClass}"
+                    registeredToString(it)
+                }})"
+            )
+            if (includeOwnerClass) {
+                sb.append(", from ${traceEvent.ownerClass}")
+            }
+            if (logStackTrace && !traceEvent.args.isEmpty()) {
+                val lastArg = traceEvent.args.last()
+                if (lastArg != null && lastArg is Throwable) {
+                    sb.append("\n")
+                    sb.append(formatThrowable(lastArg, logStackTrace))
+                }
```

Comment:
Redundant null-check? 

Alternatively:
```suggestion
                (traceEvent.args.last() as? Throwable)?.let {
                    sb.append("\n")
                    sb.append(formatThrowable(it, logStackTrace))
                }
```